From 3d5cd6c3620bbcbd4465b028027989cc87fd7c3f Mon Sep 17 00:00:00 2001
From: sanchuanhehe <wyihe5220@gmail.com>
Date: Fri, 16 Jan 2026 20:40:14 +0800
Subject: [PATCH] Add OpenHarmony (OHOS) target support to GCC

This patch adds comprehensive support for OpenHarmony OS across multiple
architectures (AArch64, ARM, x86/x86_64, RISC-V, MIPS). OpenHarmony is an
open-source distributed operating system that uses musl libc.

The implementation follows similar patterns to existing Linux targets,
with architecture-specific headers defining dynamic linker paths, library
directories, and compilation flags. Key features include:

- OHOS-specific target recognition in config.sub
- Multi-architecture support with proper triplet definitions
- Dynamic linker paths following musl conventions
- Architecture-specific library and header search paths
- Thread support via POSIX threads (similar to other Linux targets)
- Stack protection enabled by default for security

ChangeLog:

* config.sub: Add OHOS target recognition in OS validation
and kernel-OS combinations for linux-ohos triplets.

gcc/ChangeLog:

* config.gcc: Add OHOS target cases for aarch64, arm, i386/x86_64,
riscv, and mips architectures. Set DEFAULT_LIBC to LIBC_MUSL for
OHOS targets. Add OHOS to indirect function support list.
* config/aarch64/aarch64-ohos.h: New file. AArch64-specific OHOS
configuration with LP64 ABI support and Cortex-A53 errata fixes.
* config/aarch64/t-aarch64-ohos: New file. AArch64 OHOS makefile
fragment with multilib configuration.
* config/arm/arm-ohos.h: New file. ARM-specific OHOS configuration
with EABI support and float ABI variants (soft/hard float).
* config/arm/t-arm-ohos: New file. ARM OHOS makefile fragment.
* config/i386/ohos.h: New file. i386-specific OHOS configuration
for 32-bit x86 targets.
* config/i386/ohos64.h: New file. x86_64-specific OHOS configuration
for 64-bit x86 targets.
* config/mips/mips-ohos.h: New file. MIPS-specific OHOS configuration
supporting both big-endian and little-endian variants.
* config/ohos.h: New file. Generic OHOS configuration with musl libc
integration, security features, and library specifications.
* config/riscv/riscv-ohos.h: New file. RISC-V-specific OHOS
configuration with relaxation support.
* config/t-ohos: New file. OHOS-specific makefile fragment adjusting
multilib directory names from linux-gnu to linux-ohos.

libgcc/ChangeLog:

* config.host: Add OHOS target configurations for aarch64, arm,
i386/x86_64, riscv, and mips architectures with appropriate
unwind and atomic support.
* config/t-ohos-crtstuff: New file. OHOS-specific CRT object
compilation flags.
* gthr-posix.h: Exclude pthread_cancel on OHOS targets
(similar to Bionic) as it is not supported by musl libc in OHOS.

libstdc++-v3/ChangeLog:

* configure.host: Add linux-ohos case to use generic OS headers,
consistent with musl-based Linux distributions.
---
 config.sub                        |   4 +-
 gcc/config.gcc                    |  84 ++++++++++++++++-
 gcc/config/aarch64/aarch64-ohos.h | 138 ++++++++++++++++++++++++++++
 gcc/config/aarch64/t-aarch64-ohos |  28 ++++++
 gcc/config/arm/arm-ohos.h         | 121 ++++++++++++++++++++++++
 gcc/config/arm/t-arm-ohos         |  28 ++++++
 gcc/config/i386/ohos.h            | 115 +++++++++++++++++++++++
 gcc/config/i386/ohos64.h          | 117 +++++++++++++++++++++++
 gcc/config/mips/mips-ohos.h       | 129 ++++++++++++++++++++++++++
 gcc/config/ohos.h                 | 148 ++++++++++++++++++++++++++++++
 gcc/config/riscv/riscv-ohos.h     | 118 ++++++++++++++++++++++++
 gcc/config/t-ohos                 |   2 +
 libgcc/config.host                |  52 ++++++++++-
 libgcc/config/t-ohos-crtstuff     |   4 +
 libgcc/gthr-posix.h               |   4 +-
 libstdc++-v3/configure.host       |   3 +
 16 files changed, 1089 insertions(+), 6 deletions(-)
 create mode 100644 gcc/config/aarch64/aarch64-ohos.h
 create mode 100644 gcc/config/aarch64/t-aarch64-ohos
 create mode 100644 gcc/config/arm/arm-ohos.h
 create mode 100644 gcc/config/arm/t-arm-ohos
 create mode 100644 gcc/config/i386/ohos.h
 create mode 100644 gcc/config/i386/ohos64.h
 create mode 100644 gcc/config/mips/mips-ohos.h
 create mode 100644 gcc/config/ohos.h
 create mode 100644 gcc/config/riscv/riscv-ohos.h
 create mode 100644 gcc/config/t-ohos
 create mode 100644 libgcc/config/t-ohos-crtstuff

diff --git a/config.sub b/config.sub
index 38f3d037a..0a04e2db3 100755
--- a/config.sub
+++ b/config.sub
@@ -1749,7 +1749,7 @@ case $os in
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
 	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx* | zephyr* \
-	     | fiwix* )
+	     | fiwix* | ohos* )
 		;;
 	# This one is extra strict with allowed versions
 	sco3.2v2 | sco3.2v[4-9]* | sco5v6*)
@@ -1767,7 +1767,7 @@ esac
 # (given a valid OS), if there is a kernel.
 case $kernel-$os in
 	linux-gnu* | linux-dietlibc* | linux-android* | linux-newlib* \
-		   | linux-musl* | linux-relibc* | linux-uclibc* )
+		   | linux-musl* | linux-relibc* | linux-uclibc* | linux-ohos* )
 		;;
 	uclinux-uclibc* )
 		;;
diff --git a/gcc/config.gcc b/gcc/config.gcc
index 087aaa7be..0e94e6fab 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -920,6 +920,9 @@ case ${target} in
     *-*-*musl*)
       tm_defines="$tm_defines DEFAULT_LIBC=LIBC_MUSL"
       ;;
+    *-*-*ohos*)
+      tm_defines="$tm_defines DEFAULT_LIBC=LIBC_MUSL"
+      ;;
     *)
       tm_defines="$tm_defines DEFAULT_LIBC=LIBC_GLIBC"
       ;;
@@ -1236,6 +1239,81 @@ aarch64*-*-netbsd*)
 	tmake_file="${tmake_file} aarch64/t-aarch64 aarch64/t-aarch64-netbsd"
 	extra_options="${extra_options} netbsd.opt netbsd-elf.opt"
 	;;
+aarch64*-*-linux-ohos*)
+	tm_file="${tm_file} elfos.h gnu-user.h linux.h ohos.h glibc-stdint.h"
+	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-errata.h aarch64/aarch64-linux.h aarch64/aarch64-ohos.h"
+	tmake_file="${tmake_file} aarch64/t-aarch64 aarch64/t-aarch64-ohos"
+	tm_defines="${tm_defines} TARGET_DEFAULT_ASYNC_UNWIND_TABLES=1"
+	gas=yes gnu_ld=yes
+	;;
+arm*-*-linux-ohos*)
+	tm_file="elfos.h gnu-user.h linux.h ohos.h glibc-stdint.h arm/elf.h arm/linux-gas.h arm/linux-elf.h"
+	case $target in
+	arm*b-*-linux-ohos*)
+	    tm_defines="${tm_defines} TARGET_BIG_ENDIAN_DEFAULT=1"
+	    ;;
+	esac
+	tmake_file="${tmake_file} arm/t-arm arm/t-arm-elf arm/t-bpabi arm/t-linux-eabi arm/t-arm-ohos"
+	tm_file="$tm_file arm/bpabi.h arm/linux-eabi.h arm/aout.h arm/arm.h arm/arm-ohos.h"
+	gas=yes gnu_ld=yes
+	;;
+i[34567]86-*-linux-ohos* | x86_64-*-linux-ohos*)
+	tm_file="${tm_file} i386/unix.h i386/att.h elfos.h gnu-user.h ohos.h glibc-stdint.h \
+		 i386/gnu-user-common.h"
+	case ${target} in
+	i[34567]86-*-linux-ohos*)
+		tm_file="${tm_file} linux.h i386/linux-common.h i386/linux.h i386/ohos.h linux-android.h"
+		;;
+	x86_64-*-linux-ohos*)
+		tm_file="${tm_file} i386/x86-64.h i386/gnu-user64.h linux.h i386/linux-common.h i386/linux64.h i386/ohos64.h linux-android.h"
+		;;
+	esac
+	gas=yes gnu_ld=yes
+	;;
+riscv*-*-linux-ohos*)
+	tm_file="elfos.h gnu-user.h linux.h ohos.h glibc-stdint.h ${tm_file} riscv/linux.h riscv/riscv-ohos.h"
+	case "x${enable_multilib}" in
+	xno) ;;
+	xyes) tmake_file="${tmake_file} riscv/t-linux-multilib" ;;
+	*) echo "Unknown value for enable_multilib"; exit 1
+	esac
+	tmake_file="${tmake_file} riscv/t-riscv riscv/t-linux"
+	tm_defines="${tm_defines} TARGET_DEFAULT_ASYNC_UNWIND_TABLES=1"
+	gas=yes gnu_ld=yes
+	;;
+mips*-*-linux-ohos*)
+	tm_file="elfos.h gnu-user.h linux.h ohos.h glibc-stdint.h ${tm_file} mips/linux.h mips/mips-ohos.h"
+	tmake_file="${tmake_file} mips/t-mips"
+	extra_options="${extra_options} linux.opt"
+	case ${target} in
+	mipsisa32r6*-*-linux-ohos*)
+		default_mips_arch=mips32r6
+		;;
+	mipsisa32r2*-*-linux-ohos*)
+		default_mips_arch=mips32r2
+		;;
+	mipsisa32*-*-linux-ohos*)
+		default_mips_arch=mips32
+		;;
+	mips64octeon*-*-linux-ohos*)
+		default_mips_abi=n32
+		default_mips_arch=octeon
+		;;
+	mipsisa64r6*-*-linux-ohos*)
+		default_mips_abi=n32
+		default_mips_arch=mips64r6
+		;;
+	mipsisa64r2*-*-linux-ohos*)
+		default_mips_abi=n32
+		default_mips_arch=mips64r2
+		;;
+	mips64*-*-linux-ohos* | mipsisa64*-*-linux-ohos*)
+		default_mips_abi=n32
+		default_mips_arch=mips64
+		;;
+	esac
+	gas=yes gnu_ld=yes
+	;;
 aarch64*-*-linux*)
 	tm_file="${tm_file} elfos.h gnu-user.h linux.h glibc-stdint.h"
 	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-errata.h aarch64/aarch64-linux.h"
@@ -3660,6 +3738,9 @@ case ${target} in
 *-linux-musl*)
 	tmake_file="${tmake_file} t-musl"
 	;;
+*-linux-ohos*)
+	tmake_file="${tmake_file} t-ohos"
+	;;
 *-linux-uclibc*)
 	tmake_file="${tmake_file} t-uclibc"
 	;;
@@ -3668,7 +3749,7 @@ esac
 # Assume the existence of indirect function support and allow the use of the
 # resolver attribute.
 case ${target} in
-*-*-linux*android*|*-*-linux*uclibc*|*-*-linux*musl*)
+*-*-linux*android*|*-*-linux*uclibc*|*-*-linux*musl*|*-*-linux*ohos*)
         ;;
 *-*-kfreebsd*-gnu | *-*-kopensolaris*-gnu)
         ;;
@@ -5108,6 +5189,7 @@ case "${target}" in
 		case ${target} in
 		  *-linux-gnu*) triplet_os="linux-gnu";;
 		  *-linux-musl*) triplet_os="linux-musl";;
+		  *-linux-ohos*) triplet_os="linux-ohos";;
 		  *-elf*) triplet_os="elf";;
 		  *)
 			  echo "Unsupported target ${target}." 1>&2
diff --git a/gcc/config/aarch64/aarch64-ohos.h b/gcc/config/aarch64/aarch64-ohos.h
new file mode 100644
index 000000000..94b17a59c
--- /dev/null
+++ b/gcc/config/aarch64/aarch64-ohos.h
@@ -0,0 +1,138 @@
+/* Definitions for AArch64 running OpenHarmony (OHOS).
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+   This file is part of GCC.
+
+   GCC is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3, or (at your option)
+   any later version.
+
+   GCC is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GCC; see the file COPYING3.  If not see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef GCC_AARCH64_OHOS_H
+#define GCC_AARCH64_OHOS_H
+
+/* Musl libc in OpenHarmony provides SSP support (stack_chk_fail) */
+#define TARGET_LIBC_PROVIDES_SSP 1
+
+/* Define architecture-specific dynamic linker for AArch64 */
+#define OHOS_ARCH_DYNAMIC_LINKER "/lib/ld-musl-aarch64.so.1"
+
+/* Include the generic OpenHarmony configuration */
+/* Note: ohos.h is already included via tm.h, no need to include again */
+
+/* AArch64 OpenHarmony uses the LP64 ABI */
+#undef TARGET_FIX_ERR_A53_835769_DEFAULT
+#define TARGET_FIX_ERR_A53_835769_DEFAULT 1
+#undef TARGET_FIX_ERR_A53_843419_DEFAULT
+#define TARGET_FIX_ERR_A53_843419_DEFAULT 1
+
+/* OpenHarmony on AArch64 uses musl-style 64-bit dynamic linker.
+   The dynamic linker is already selected by architecture-aware logic
+   in ohos.h, so we don't need to override it here unless there are
+   AArch64-specific requirements. */
+
+/* AArch64-specific OpenHarmony defines - override Linux configuration */
+#ifdef TARGET_OS_CPP_BUILTINS
+#undef TARGET_OS_CPP_BUILTINS
+#endif
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      GNU_USER_TARGET_OS_CPP_BUILTINS();	\
+      OHOS_BASE_OS_CPP_BUILTINS();		\
+      builtin_define ("__aarch64__");		\
+      builtin_assert ("cpu=aarch64");		\
+      builtin_assert ("machine=aarch64");	\
+    }						\
+  while (0)
+
+/* OpenHarmony library paths for AArch64 */
+#define OHOS_ARCH_LIB_DIR "aarch64-linux-ohos"
+
+/* Inject architecture-specific include directories via the CPP spec. */
+#define OHOS_MULTIARCH_CPP_SPEC OHOS_MULTIARCH_INCLUDE_SPEC(OHOS_ARCH_LIB_DIR)
+#ifndef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC ""
+#endif
+#define OHOS_PREV_SUBTARGET_CPP_SPEC SUBTARGET_CPP_SPEC
+#undef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC OHOS_PREV_SUBTARGET_CPP_SPEC OHOS_MULTIARCH_CPP_SPEC
+#undef OHOS_PREV_SUBTARGET_CPP_SPEC
+#undef OHOS_MULTIARCH_CPP_SPEC
+
+/* Enhanced library search paths for OpenHarmony */
+#undef STANDARD_STARTFILE_PREFIX_1
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+
+#undef STANDARD_STARTFILE_PREFIX_2
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Library directories for OHOS - use flat structure without arch subdirs */
+#undef MD_EXEC_PREFIX
+#define MD_EXEC_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX
+#define MD_STARTFILE_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX_1
+#define MD_STARTFILE_PREFIX_1 "/lib/"
+
+/* Override STARTFILE and ENDFILE specs - OHOS uses flat /usr/lib/ structure */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:%R/usr/lib/crt1.o%s; \
+       pie:%R/usr/lib/Scrt1.o%s; \
+       :%R/usr/lib/crt1.o%s}} \
+   %R/usr/lib/crti.o%s \
+   %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} \
+   %R/usr/lib/crtn.o%s"
+
+#ifdef LINK_SPEC
+#undef LINK_SPEC
+#endif
+/* Define the link spec for AArch64 OpenHarmony */
+#define LINK_SPEC						\
+  "%{mbig-endian:-EB} %{mlittle-endian:-EL} "			\
+  "%{shared:-shared} "						\
+  "%{!shared: "							\
+    "%{!static: "						\
+      "%{rdynamic:-export-dynamic} "				\
+      "-dynamic-linker " OHOS_DYNAMIC_LINKER "} "		\
+    "%{static:-static}} "					\
+  "%{mfix-cortex-a53-835769:--fix-cortex-a53-835769} "		\
+  "%{mfix-cortex-a53-843419:--fix-cortex-a53-843419} "		\
+  "-L%R/usr/lib " 						\
+  "-L%R/lib "
+
+/* Default to little endian */
+#undef TARGET_DEFAULT_ASYNC_UNWIND_TABLES
+#define TARGET_DEFAULT_ASYNC_UNWIND_TABLES 1
+
+/* AArch64 OpenHarmony multilib configuration */
+#undef MULTILIB_DEFAULTS
+#define MULTILIB_DEFAULTS { "mabi=lp64" }
+
+/* Use the AAPCS64 calling convention */
+#undef TARGET_DEFAULT
+#define TARGET_DEFAULT (MASK_LITTLE_ENDIAN)
+
+/* Support for outline atomics */
+#ifndef TARGET_OUTLINE_ATOMICS
+#define TARGET_OUTLINE_ATOMICS 1
+#endif
+
+#endif /* GCC_AARCH64_OHOS_H */
diff --git a/gcc/config/aarch64/t-aarch64-ohos b/gcc/config/aarch64/t-aarch64-ohos
new file mode 100644
index 000000000..ed0fb53ee
--- /dev/null
+++ b/gcc/config/aarch64/t-aarch64-ohos
@@ -0,0 +1,28 @@
+# Machine description for AArch64 architecture - OpenHarmony (OHOS) specific.
+#  Copyright (C) 2026 Free Software Foundation, Inc.
+#  Contributed by ARM Ltd.
+#
+#  This file is part of GCC.
+#
+#  GCC is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3, or (at your option)
+#  any later version.
+#
+#  GCC is distributed in the hope that it will be useful, but
+#  WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with GCC; see the file COPYING3.  If not see
+#  <http://www.gnu.org/licenses/>.
+
+LIB1ASMSRC   = aarch64/lib1funcs.asm
+LIB1ASMFUNCS = _aarch64_sync_cache_range
+
+AARCH_BE = $(if $(findstring TARGET_BIG_ENDIAN_DEFAULT=1, $(tm_defines)),_be)
+MULTILIB_OSDIRNAMES = mabi.lp64=../lib64$(call if_multiarch,:aarch64$(AARCH_BE)-linux-ohos)
+MULTIARCH_DIRNAME = $(call if_multiarch,aarch64$(AARCH_BE)-linux-ohos)
+
+MULTILIB_OSDIRNAMES += mabi.ilp32=../libilp32$(call if_multiarch,:aarch64$(AARCH_BE)-linux-ohos_ilp32)
diff --git a/gcc/config/arm/arm-ohos.h b/gcc/config/arm/arm-ohos.h
new file mode 100644
index 000000000..ac3127b29
--- /dev/null
+++ b/gcc/config/arm/arm-ohos.h
@@ -0,0 +1,121 @@
+/* Definitions for ARM running OpenHarmony (OHOS).
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+   This file is part of GCC.
+
+   GCC is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3, or (at your option)
+   any later version.
+
+   GCC is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GCC; see the file COPYING3.  If not see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef GCC_ARM_OHOS_H
+#define GCC_ARM_OHOS_H
+
+/* Define architecture-specific dynamic linker with float ABI support.
+   Use spec string to select at link time based on -mfloat-abi option. */
+#define OHOS_DYNAMIC_LINKER_SOFT "/lib/ld-musl-arm.so.1"
+#define OHOS_DYNAMIC_LINKER_HARD "/lib/ld-musl-armhf.so.1"
+#define OHOS_ARCH_DYNAMIC_LINKER \
+  "%{mfloat-abi=hard:" OHOS_DYNAMIC_LINKER_HARD ";:" OHOS_DYNAMIC_LINKER_SOFT "}"
+
+/* Include the generic OpenHarmony configuration */
+/* Note: ohos.h is already included via tm.h, no need to include again */
+
+/* ARM-specific OpenHarmony defines - override Linux configuration */
+#ifdef TARGET_OS_CPP_BUILTINS
+#undef TARGET_OS_CPP_BUILTINS
+#endif
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      GNU_USER_TARGET_OS_CPP_BUILTINS();	\
+      OHOS_BASE_OS_CPP_BUILTINS();		\
+      builtin_define ("__arm__");		\
+      builtin_assert ("cpu=arm");		\
+      builtin_assert ("machine=arm");		\
+    }						\
+  while (0)
+
+/* OpenHarmony library paths for ARM - use consistent naming */
+#define OHOS_ARCH_LIB_DIR "arm-linux-ohos"
+
+/* Inject architecture-specific include directories via the CPP spec. */
+#define OHOS_MULTIARCH_CPP_SPEC OHOS_MULTIARCH_INCLUDE_SPEC(OHOS_ARCH_LIB_DIR)
+#ifndef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC ""
+#endif
+#define OHOS_PREV_SUBTARGET_CPP_SPEC SUBTARGET_CPP_SPEC
+#undef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC OHOS_PREV_SUBTARGET_CPP_SPEC OHOS_MULTIARCH_CPP_SPEC
+#undef OHOS_PREV_SUBTARGET_CPP_SPEC
+#undef OHOS_MULTIARCH_CPP_SPEC
+
+/* Enhanced library search paths for OpenHarmony */
+#undef STANDARD_STARTFILE_PREFIX_1
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+
+#undef STANDARD_STARTFILE_PREFIX_2
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Library directories for OHOS - use flat structure without arch subdirs */
+#undef MD_EXEC_PREFIX
+#define MD_EXEC_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX
+#define MD_STARTFILE_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX_1
+#define MD_STARTFILE_PREFIX_1 "/lib/"
+
+/* Override STARTFILE and ENDFILE specs - OHOS uses flat /usr/lib/ structure */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:%R/usr/lib/crt1.o%s; \
+       pie:%R/usr/lib/Scrt1.o%s; \
+       :%R/usr/lib/crt1.o%s}} \
+   %R/usr/lib/crti.o%s \
+   %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} \
+   %R/usr/lib/crtn.o%s"
+#ifdef LINK_SPEC
+#undef LINK_SPEC
+#endif
+/* Define the link spec for ARM OpenHarmony */
+#define LINK_SPEC						\
+  "%{mbig-endian:-EB} %{mlittle-endian:-EL} "			\
+  "%{shared:-shared} "						\
+  "%{!shared: "							\
+    "%{!static: "						\
+      "%{rdynamic:-export-dynamic} "				\
+      "-dynamic-linker " OHOS_DYNAMIC_LINKER "} "		\
+    "%{static:-static}} "					\
+  "-L%R/usr/lib " 						\
+  "-L%R/lib "
+
+/* ARM OpenHarmony uses EABI by default */
+#undef TARGET_DEFAULT_FLOAT_ABI
+#define TARGET_DEFAULT_FLOAT_ABI ARM_FLOAT_ABI_SOFT
+
+/* Default to little endian and async unwind tables */
+#undef TARGET_DEFAULT_ASYNC_UNWIND_TABLES
+#define TARGET_DEFAULT_ASYNC_UNWIND_TABLES 1
+
+/* ARM OpenHarmony multilib configuration */
+#undef MULTILIB_DEFAULTS
+#define MULTILIB_DEFAULTS \
+  { "marm", "mlittle-endian", "mfloat-abi=soft", "mno-thumb-interwork" }
+
+#endif /* GCC_ARM_OHOS_H */
diff --git a/gcc/config/arm/t-arm-ohos b/gcc/config/arm/t-arm-ohos
new file mode 100644
index 000000000..23ed89983
--- /dev/null
+++ b/gcc/config/arm/t-arm-ohos
@@ -0,0 +1,28 @@
+# Machine description for ARM architecture - OpenHarmony (OHOS) specific.
+#  Copyright (C) 2026 Free Software Foundation, Inc.
+#  Contributed by ARM Ltd.
+#
+#  This file is part of GCC.
+#
+#  GCC is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3, or (at your option)
+#  any later version.
+#
+#  GCC is distributed in the hope that it will be useful, but
+#  WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with GCC; see the file COPYING3.  If not see
+#  <http://www.gnu.org/licenses/>.
+
+LIB1ASMSRC = arm/lib1funcs.S
+LIB1ASMFUNCS = _dvmd_tls _bb_init_func
+
+ARM_BE = $(if $(findstring TARGET_BIG_ENDIAN_DEFAULT=1, $(tm_defines)),_be)
+MULTILIB_OSDIRNAMES = mfloat-abi.soft=../lib$(call if_multiarch,:arm$(ARM_BE)-linux-ohos)
+MULTIARCH_DIRNAME = $(call if_multiarch,arm$(ARM_BE)-linux-ohos)
+
+MULTILIB_OSDIRNAMES += mfloat-abi.hard=../lib/arm$(ARM_BE)-linux-ohoshf$(call if_multiarch,:arm$(ARM_BE)-linux-ohoshf)
diff --git a/gcc/config/i386/ohos.h b/gcc/config/i386/ohos.h
new file mode 100644
index 000000000..c5e737055
--- /dev/null
+++ b/gcc/config/i386/ohos.h
@@ -0,0 +1,115 @@
+/* Definitions for i386 running OpenHarmony (OHOS).
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+   This file is part of GCC.
+
+   GCC is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3, or (at your option)
+   any later version.
+
+   GCC is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GCC; see the file COPYING3.  If not see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef GCC_I386_OHOS_H
+#define GCC_I386_OHOS_H
+
+/* Define architecture-specific dynamic linker - match LLVM getDynamicLinker() */
+#define OHOS_ARCH_DYNAMIC_LINKER "/lib/ld-musl-i386.so.1"
+
+
+/* Include the generic OpenHarmony configuration */
+/* Note: ohos.h is already included via tm.h, no need to include again */
+
+/* x86/x86_64-specific OpenHarmony defines - override Linux configuration */
+#ifdef TARGET_OS_CPP_BUILTINS
+#undef TARGET_OS_CPP_BUILTINS
+#endif
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      /* Include base GNU and OHOS defines */ \
+      GNU_USER_TARGET_OS_CPP_BUILTINS ();	\
+      OHOS_BASE_OS_CPP_BUILTINS ();		\
+      /* x86-specific defines already in GNU_USER_TARGET_OS_CPP_BUILTINS */ \
+    }						\
+  while (0)
+
+/* OpenHarmony library paths for i386 */
+
+# define OHOS_ARCH_LIB_DIR "i386-linux-ohos"
+
+
+/* Inject architecture-specific include directories via the CPP spec. */
+#define OHOS_MULTIARCH_CPP_SPEC OHOS_MULTIARCH_INCLUDE_SPEC(OHOS_ARCH_LIB_DIR)
+#ifndef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC ""
+#endif
+#define OHOS_PREV_SUBTARGET_CPP_SPEC SUBTARGET_CPP_SPEC
+#undef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC OHOS_PREV_SUBTARGET_CPP_SPEC OHOS_MULTIARCH_CPP_SPEC
+#undef OHOS_PREV_SUBTARGET_CPP_SPEC
+#undef OHOS_MULTIARCH_CPP_SPEC
+#undef STANDARD_STARTFILE_PREFIX_1
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+
+#undef STANDARD_STARTFILE_PREFIX_2
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Library directories for OHOS - use flat structure without arch subdirs */
+#undef MD_EXEC_PREFIX
+#define MD_EXEC_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX
+#define MD_STARTFILE_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX_1
+#define MD_STARTFILE_PREFIX_1 "/lib/"
+
+/* Override STARTFILE and ENDFILE specs - OHOS uses flat /usr/lib/ structure */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:%R/usr/lib/crt1.o%s; \
+       pie:%R/usr/lib/Scrt1.o%s; \
+       :%R/usr/lib/crt1.o%s}} \
+   %R/usr/lib/crti.o%s \
+   %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} \
+   %R/usr/lib/crtn.o%s"
+
+/* Override CC1_SPEC from linux-common.h to remove Android logic */
+#undef CC1_SPEC
+#define CC1_SPEC GNU_USER_TARGET_CC1_SPEC
+
+/* Override LIB_SPEC from linux-common.h to remove Android logic */
+#undef LIB_SPEC
+#define LIB_SPEC GNU_USER_TARGET_LIB_SPEC
+
+#undef LINK_SPEC
+/* Define the link spec for i386 OpenHarmony - 32-bit only, no multilib */
+#define LINK_SPEC						\
+  "-m elf_i386 "						\
+  "%{shared:-shared} "						\
+  "%{!shared: "							\
+    "%{!static: "						\
+      "%{rdynamic:-export-dynamic} "				\
+      "-dynamic-linker " OHOS_DYNAMIC_LINKER "} "		\
+    "%{static:-static}} "					\
+  "-L%R/usr/lib " 						\
+  "-L%R/lib "
+
+/* Default to little endian and async unwind tables */
+#undef TARGET_DEFAULT_ASYNC_UNWIND_TABLES
+#define TARGET_DEFAULT_ASYNC_UNWIND_TABLES 1
+
+#endif /* GCC_I386_OHOS_H */
diff --git a/gcc/config/i386/ohos64.h b/gcc/config/i386/ohos64.h
new file mode 100644
index 000000000..a09bd2603
--- /dev/null
+++ b/gcc/config/i386/ohos64.h
@@ -0,0 +1,117 @@
+/* Definitions for x86_64 running OpenHarmony (OHOS).
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+   This file is part of GCC.
+
+   GCC is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3, or (at your option)
+   any later version.
+
+   GCC is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GCC; see the file COPYING3.  If not see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef GCC_I386_OHOS64_H
+#define GCC_I386_OHOS64_H
+
+/* Define architecture-specific dynamic linker - match LLVM getDynamicLinker() */
+#define OHOS_ARCH_DYNAMIC_LINKER "/lib/ld-musl-x86_64.so.1"
+
+
+/* Include the generic OpenHarmony configuration */
+/* Note: ohos.h is already included via tm.h, no need to include again */
+
+/* x86/x86_64-specific OpenHarmony defines - override Linux configuration */
+#ifdef TARGET_OS_CPP_BUILTINS
+#undef TARGET_OS_CPP_BUILTINS
+#endif
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      /* Include base GNU and OHOS defines */ \
+      GNU_USER_TARGET_OS_CPP_BUILTINS ();	\
+      OHOS_BASE_OS_CPP_BUILTINS ();		\
+      /* x86-specific defines already in GNU_USER_TARGET_OS_CPP_BUILTINS */ \
+    }						\
+  while (0)
+
+/* OpenHarmony library paths for x86 */
+
+# define OHOS_ARCH_LIB_DIR "x86_64-linux-ohos"
+
+
+/* Inject architecture-specific include directories via the CPP spec. */
+#define OHOS_MULTIARCH_CPP_SPEC OHOS_MULTIARCH_INCLUDE_SPEC(OHOS_ARCH_LIB_DIR)
+#ifndef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC ""
+#endif
+#define OHOS_PREV_SUBTARGET_CPP_SPEC SUBTARGET_CPP_SPEC
+#undef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC OHOS_PREV_SUBTARGET_CPP_SPEC OHOS_MULTIARCH_CPP_SPEC
+#undef OHOS_PREV_SUBTARGET_CPP_SPEC
+#undef OHOS_MULTIARCH_CPP_SPEC
+#undef STANDARD_STARTFILE_PREFIX_1
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+
+#undef STANDARD_STARTFILE_PREFIX_2
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Library directories for OHOS - use flat structure without arch subdirs */
+#undef MD_EXEC_PREFIX
+#define MD_EXEC_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX
+#define MD_STARTFILE_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX_1
+#define MD_STARTFILE_PREFIX_1 "/lib/"
+
+/* Override STARTFILE and ENDFILE specs - OHOS uses flat /usr/lib/ structure */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:%R/usr/lib/crt1.o%s; \
+       pie:%R/usr/lib/Scrt1.o%s; \
+       :%R/usr/lib/crt1.o%s}} \
+   %R/usr/lib/crti.o%s \
+   %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} \
+   %R/usr/lib/crtn.o%s"
+
+/* Override CC1_SPEC from linux-common.h to remove Android logic */
+#undef CC1_SPEC
+#define CC1_SPEC GNU_USER_TARGET_CC1_SPEC
+
+/* Override LIB_SPEC from linux-common.h to remove Android logic */
+#undef LIB_SPEC
+#define LIB_SPEC GNU_USER_TARGET_LIB_SPEC
+
+#undef LINK_SPEC
+/* Define the link spec for x86 OpenHarmony */
+#define LINK_SPEC						\
+  "%{" SPEC_32 ":-m elf_i386} "				\
+  "%{" SPEC_64 ":-m elf_x86_64} "				\
+  "%{shared:-shared} "						\
+  "%{!shared: "							\
+    "%{!static: "						\
+      "%{rdynamic:-export-dynamic} "				\
+      "%{" SPEC_32 ":-dynamic-linker " OHOS_DYNAMIC_LINKER "} "	\
+      "%{" SPEC_64 ":-dynamic-linker " OHOS_DYNAMIC_LINKER "}} "	\
+    "%{static:-static}} "					\
+  "-L%R/usr/lib " 						\
+  "-L%R/lib "
+
+/* Default to little endian and async unwind tables */
+#undef TARGET_DEFAULT_ASYNC_UNWIND_TABLES
+#define TARGET_DEFAULT_ASYNC_UNWIND_TABLES 1
+
+#endif /* GCC_I386_OHOS64_H */
diff --git a/gcc/config/mips/mips-ohos.h b/gcc/config/mips/mips-ohos.h
new file mode 100644
index 000000000..cc0cd38ea
--- /dev/null
+++ b/gcc/config/mips/mips-ohos.h
@@ -0,0 +1,129 @@
+/* Definitions for MIPS running OpenHarmony (OHOS).
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+   This file is part of GCC.
+
+   GCC is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3, or (at your option)
+   any later version.
+
+   GCC is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GCC; see the file COPYING3.  If not see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef GCC_MIPS_OHOS_H
+#define GCC_MIPS_OHOS_H
+
+/* Define architecture-specific dynamic linker - use spec strings for runtime selection.
+   Match LLVM getDynamicLinker() and musl conventions. */
+#define OHOS_DYNAMIC_LINKER32 \
+  "/lib/ld-musl-mips%{mips32r6|mips64r6:r6}%{EL:el}%{msoft-float:-sf}.so.1"
+#define OHOS_DYNAMIC_LINKER64 \
+  "/lib/ld-musl-mips64%{mips64r6:r6}%{EL:el}%{msoft-float:-sf}.so.1"
+#define OHOS_DYNAMIC_LINKERN32 \
+  "/lib/ld-musl-mipsn32%{mips64r6:r6}%{EL:el}%{msoft-float:-sf}.so.1"
+
+/* Select appropriate linker based on ABI - use same pattern as gnu-user.h */
+#define OHOS_ARCH_DYNAMIC_LINKER32 OHOS_DYNAMIC_LINKER32
+#define OHOS_ARCH_DYNAMIC_LINKER64 OHOS_DYNAMIC_LINKER64
+#define OHOS_ARCH_DYNAMIC_LINKERN32 OHOS_DYNAMIC_LINKERN32
+
+/* Include the generic OpenHarmony configuration */
+/* Note: ohos.h is already included via tm.h, no need to include again */
+
+/* MIPS-specific OpenHarmony defines - override Linux configuration */
+#ifdef TARGET_OS_CPP_BUILTINS
+#undef TARGET_OS_CPP_BUILTINS
+#endif
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      /* Include base GNU and OHOS defines */ \
+      GNU_USER_TARGET_OS_CPP_BUILTINS ();	\
+      OHOS_BASE_OS_CPP_BUILTINS ();		\
+      /* MIPS-specific defines */ \
+      builtin_define ("__mips__");		\
+      builtin_assert ("cpu=mips");		\
+      builtin_assert ("machine=mips");		\
+    }						\
+  while (0)
+
+/* OpenHarmony library paths for MIPS - use consistent naming */
+#define OHOS_ARCH_LIB_DIR "mips-linux-ohos"
+
+/* Inject architecture-specific include directories via the CPP spec. */
+#define OHOS_MULTIARCH_CPP_SPEC OHOS_MULTIARCH_INCLUDE_SPEC(OHOS_ARCH_LIB_DIR)
+#ifndef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC ""
+#endif
+#define OHOS_PREV_SUBTARGET_CPP_SPEC SUBTARGET_CPP_SPEC
+#undef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC OHOS_PREV_SUBTARGET_CPP_SPEC OHOS_MULTIARCH_CPP_SPEC
+#undef OHOS_PREV_SUBTARGET_CPP_SPEC
+#undef OHOS_MULTIARCH_CPP_SPEC
+
+/* Enhanced library search paths for OpenHarmony */
+#undef STANDARD_STARTFILE_PREFIX_1
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+
+#undef STANDARD_STARTFILE_PREFIX_2
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Library directories for OHOS - use flat structure without arch subdirs */
+#undef MD_EXEC_PREFIX
+#define MD_EXEC_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX
+#define MD_STARTFILE_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX_1
+#define MD_STARTFILE_PREFIX_1 "/lib/"
+
+/* Override STARTFILE and ENDFILE specs - OHOS uses flat /usr/lib/ structure */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:%R/usr/lib/crt1.o%s; \
+       pie:%R/usr/lib/Scrt1.o%s; \
+       :%R/usr/lib/crt1.o%s}} \
+   %R/usr/lib/crti.o%s \
+   %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} \
+   %R/usr/lib/crtn.o%s"
+
+/* Linker emulation modes - same as linux.h */
+#define OHOS_LINK_EMULATION32 "elf32%{EB:b}%{EL:l}tsmip"
+#define OHOS_LINK_EMULATION64 "elf64%{EB:b}%{EL:l}tsmip"
+#define OHOS_LINK_EMULATIONN32 "elf32%{EB:b}%{EL:l}tsmipn32"
+
+/* Define the link spec for MIPS OpenHarmony - follow gnu-user.h pattern */
+#undef LINK_SPEC
+#define LINK_SPEC						\
+  "%{G*} %{EB} %{EL} %{mips*} %{shared} "			\
+  "%{!shared: "							\
+    "%{!static: "						\
+      "%{rdynamic:-export-dynamic} "				\
+      "%{mabi=n32: -dynamic-linker " OHOS_ARCH_DYNAMIC_LINKERN32 "} "	\
+      "%{mabi=64: -dynamic-linker " OHOS_ARCH_DYNAMIC_LINKER64 "} "	\
+      "%{mabi=32: -dynamic-linker " OHOS_ARCH_DYNAMIC_LINKER32 "}} "	\
+    "%{static:-static}} "					\
+  "%{mabi=n32:-m " OHOS_LINK_EMULATIONN32 "} "			\
+  "%{mabi=64:-m " OHOS_LINK_EMULATION64 "} "			\
+  "%{mabi=32:-m " OHOS_LINK_EMULATION32 "} "			\
+  "-L%R/usr/lib " 						\
+  "-L%R/lib "
+
+/* Default to little endian and async unwind tables */
+#undef TARGET_DEFAULT_ASYNC_UNWIND_TABLES
+#define TARGET_DEFAULT_ASYNC_UNWIND_TABLES 1
+
+#endif /* GCC_MIPS_OHOS_H */
diff --git a/gcc/config/ohos.h b/gcc/config/ohos.h
new file mode 100644
index 000000000..78b72bc7c
--- /dev/null
+++ b/gcc/config/ohos.h
@@ -0,0 +1,148 @@
+/* OpenHarmony (OHOS) configuration.
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GCC; see the file COPYING3.  If not see
+<http://www.gnu.org/licenses/>.  */
+
+/* Helper fragment for include dirs - OHOS uses flat /usr/include/ */
+#define OHOS_MULTIARCH_INCLUDE_SPEC(archdir) \
+  " %{!nostdinc:%{!fself-test=*: -idirafter %R/usr/include}}"
+
+/* OpenHarmony uses musl libc but with some differences */
+#undef OPTION_GLIBC
+#define OPTION_GLIBC 0
+#undef OPTION_UCLIBC
+#define OPTION_UCLIBC 0
+#undef OPTION_BIONIC
+#define OPTION_BIONIC 0
+#undef OPTION_MUSL
+#define OPTION_MUSL 1
+
+/* Base OpenHarmony target OS defines - used by architecture-specific headers */
+#define OHOS_BASE_OS_CPP_BUILTINS()				\
+  do {								\
+    builtin_define ("__OHOS__");				\
+    builtin_define ("__OpenHarmony__");				\
+    builtin_define ("__MUSL__");				\
+    builtin_assert ("target=ohos");				\
+    builtin_assert ("target=openharmony");			\
+  } while (0)
+
+
+/* OpenHarmony-specific libc selection */
+#ifndef CHOOSE_DYNAMIC_LINKER
+#define CHOOSE_DYNAMIC_LINKER(G, U, B, M, O) O
+#endif
+
+/* OpenHarmony uses musl-style dynamic linker paths similar to LLVM/Clang.
+   The dynamic linker path is constructed as "/lib/ld-musl-<arch>.so.1"
+   where <arch> is architecture-specific and may include ABI suffixes. */
+
+/* Dynamic linker configuration
+   Each architecture-specific ohos header should define OHOS_ARCH_DYNAMIC_LINKER
+   following the pattern: /lib/ld-musl-{arch}.so.1 */
+
+/* Primary dynamic linker macro - must be defined by arch-specific headers */
+#define OHOS_DYNAMIC_LINKER OHOS_ARCH_DYNAMIC_LINKER
+
+/* Standard library and runtime support - follow LLVM's Unix-style paths */
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Thread model */
+#undef TARGET_THREAD_FILE
+#define TARGET_THREAD_FILE "posix"
+
+/* Enable stack protector by default (OpenHarmony security feature) */
+#define OHOS_CC1_SPEC \
+  " %{!fno-stack-protector:%{!fstack-protector-all:%{!fstack-protector-strong:%{!fstack-protector:-fstack-protector}}}}"
+
+/* Link spec is defined by architecture-specific ohos headers */
+
+/* Default to position independent executables */
+
+#ifndef CC1PLUS_SPEC
+#define CC1PLUS_SPEC ""
+#endif
+
+#ifndef LINK_PIE_SPEC
+#ifdef HAVE_LD_PIE
+#define LINK_PIE_SPEC "%{!shared:%{!static:%{!r:-pie}}} "
+#else
+#define LINK_PIE_SPEC ""
+#endif
+#endif
+
+/* OpenHarmony sysroot structure support */
+/* Standard library paths for OHOS - use architecture-specific subdirectories */
+#define OHOS_LIB_SPEC \
+  "%{!nostdlib: \
+     %{!nodefaultlibs: \
+       %{shared:-lc} \
+       %{!shared:%{profile:-lc_p}%{!profile:-lc}} \
+       %{pthread:-lpthread}}}"
+
+/* OHOS uses architecture-specific library directories */
+#ifndef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:gcrt1.o%s;pie:Scrt1.o%s;:crt1.o%s}} \
+   crti.o%s %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+#endif
+
+#ifndef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} crtn.o%s"
+#endif
+
+/* Override LIB_SPEC to use OHOS library specification */
+#ifndef LIB_SPEC
+#define LIB_SPEC OHOS_LIB_SPEC
+#endif
+
+/* Use the same unwinding mechanism as Linux */
+/* Note: MD_UNWIND_SUPPORT is poisoned in newer GCC versions, commenting out */
+/* #define MD_UNWIND_SUPPORT "config/linux-unwind.h" */
+
+/* OpenHarmony runtime library configuration */
+/* Note: GCC uses libgcc runtime library with musl libc.
+   libgcc is fully compatible with musl - it provides low-level compiler
+   runtime support (integer arithmetic, exception handling, etc.) that
+   works independently of the C library implementation.
+   Unlike LLVM's compiler-rt, libgcc + musl is a well-established combination
+   used by many Linux distributions (Alpine, OpenWrt, etc.). */
+
+/* OpenHarmony C++ standard library configuration */
+/* Encourage libc++ usage for OHOS compatibility, but allow libstdc++ fallback */
+#ifndef CPLUSPLUS_CPP_SPEC
+#define CPLUSPLUS_CPP_SPEC "-D_GNU_SOURCE"
+#endif
+
+/* Default to libc++ for C++ programs when possible */
+#ifndef CPLUSPLUS_LIB_SPEC
+#define CPLUSPLUS_LIB_SPEC \
+  "%{!nostdlib++: \
+     %{!stdlib=libstdc++: \
+       -lc++ -lc++abi} \
+     %{stdlib=libstdc++: \
+       -lstdc++}} \
+   -lc"
+#endif
+
+/* Runtime library information */
+#define OHOS_RUNTIME_INFO \
+  "OHOS GCC toolchain: musl libc + libgcc runtime. " \
+  "This combination is fully supported and widely used in production."
diff --git a/gcc/config/riscv/riscv-ohos.h b/gcc/config/riscv/riscv-ohos.h
new file mode 100644
index 000000000..8d14f13b4
--- /dev/null
+++ b/gcc/config/riscv/riscv-ohos.h
@@ -0,0 +1,118 @@
+/* Definitions for RISC-V running OpenHarmony (OHOS).
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+   This file is part of GCC.
+
+   GCC is free software; you can redistribute it and/or modify it
+   under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3, or (at your option)
+   any later version.
+
+   GCC is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GCC; see the file COPYING3.  If not see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef GCC_RISCV_OHOS_H
+#define GCC_RISCV_OHOS_H
+
+/* Define architecture-specific dynamic linker using XLEN_SPEC for 32/64 bit.
+   XLEN_SPEC is defined in riscv.h and expands to "32" or "64" based on -march. */
+#define OHOS_ARCH_DYNAMIC_LINKER "/lib/ld-musl-riscv" XLEN_SPEC ".so.1"
+
+/* Include the generic OpenHarmony configuration */
+/* Note: ohos.h is already included via tm.h, no need to include again */
+
+/* RISC-V-specific OpenHarmony defines - override Linux configuration */
+#ifdef TARGET_OS_CPP_BUILTINS
+#undef TARGET_OS_CPP_BUILTINS
+#endif
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      /* Include base GNU and OHOS defines */ \
+      GNU_USER_TARGET_OS_CPP_BUILTINS ();	\
+      OHOS_BASE_OS_CPP_BUILTINS();		\
+      /* RISC-V specific defines */ \
+      builtin_define ("__riscv");		\
+      builtin_assert ("cpu=riscv");		\
+      builtin_assert ("machine=riscv");		\
+    }						\
+  while (0)
+
+/* OpenHarmony library paths for RISC-V - use consistent naming */
+#define OHOS_ARCH_LIB_DIR "riscv-linux-ohos"
+
+/* Inject architecture-specific include directories via the CPP spec. */
+#define OHOS_MULTIARCH_CPP_SPEC OHOS_MULTIARCH_INCLUDE_SPEC(OHOS_ARCH_LIB_DIR)
+#ifndef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC ""
+#endif
+#define OHOS_PREV_SUBTARGET_CPP_SPEC SUBTARGET_CPP_SPEC
+#undef SUBTARGET_CPP_SPEC
+#define SUBTARGET_CPP_SPEC OHOS_PREV_SUBTARGET_CPP_SPEC OHOS_MULTIARCH_CPP_SPEC
+#undef OHOS_PREV_SUBTARGET_CPP_SPEC
+#undef OHOS_MULTIARCH_CPP_SPEC
+
+/* Enhanced library search paths for OpenHarmony */
+#undef STANDARD_STARTFILE_PREFIX_1
+#define STANDARD_STARTFILE_PREFIX_1 "/usr/lib/"
+
+#undef STANDARD_STARTFILE_PREFIX_2
+#define STANDARD_STARTFILE_PREFIX_2 "/lib/"
+
+/* Library directories for OHOS - use flat structure without arch subdirs */
+#undef MD_EXEC_PREFIX
+#define MD_EXEC_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX
+#define MD_STARTFILE_PREFIX "/usr/lib/"
+
+#undef MD_STARTFILE_PREFIX_1
+#define MD_STARTFILE_PREFIX_1 "/lib/"
+
+/* Override STARTFILE and ENDFILE specs - OHOS uses flat /usr/lib/ structure */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared: \
+     %{pg|p|profile:%R/usr/lib/crt1.o%s; \
+       pie:%R/usr/lib/Scrt1.o%s; \
+       :%R/usr/lib/crt1.o%s}} \
+   %R/usr/lib/crti.o%s \
+   %{static:crtbeginT.o%s;shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC \
+  "%{shared|pie:crtendS.o%s;:crtend.o%s} \
+   %R/usr/lib/crtn.o%s"
+#ifdef LINK_SPEC
+#undef LINK_SPEC
+#endif
+/* Define the link spec for RISC-V OpenHarmony */
+#define LINK_SPEC						\
+  "%{mno-relax:--no-relax} "					\
+  "%{mbig-endian:-EB} "						\
+  "%{mlittle-endian:-EL} "					\
+  "%{shared:-shared} "						\
+  "%{!shared: "							\
+    "%{!static: "						\
+      "%{rdynamic:-export-dynamic} "				\
+      "-dynamic-linker " OHOS_DYNAMIC_LINKER "} "		\
+    "%{static:-static}} "					\
+  "-L%R/usr/lib " 						\
+  "-L%R/lib "
+
+/* RISC-V OpenHarmony uses relaxation by default */
+#ifndef TARGET_DEFAULT_RELAX
+#define TARGET_DEFAULT_RELAX 1
+#endif
+
+/* Default to async unwind tables */
+#undef TARGET_DEFAULT_ASYNC_UNWIND_TABLES
+#define TARGET_DEFAULT_ASYNC_UNWIND_TABLES 1
+
+#endif /* GCC_RISCV_OHOS_H */
diff --git a/gcc/config/t-ohos b/gcc/config/t-ohos
new file mode 100644
index 000000000..b45633a3c
--- /dev/null
+++ b/gcc/config/t-ohos
@@ -0,0 +1,2 @@
+MULTIARCH_DIRNAME := $(subst -linux-gnu,-linux-ohos,$(MULTIARCH_DIRNAME))
+MULTILIB_OSDIRNAMES := $(subst -linux-gnu,-linux-ohos,$(MULTILIB_OSDIRNAMES))
diff --git a/libgcc/config.host b/libgcc/config.host
index 6a88ee5a2..869063b9c 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -310,6 +310,11 @@ case ${host} in
   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-fuchsia"
   extra_parts="crtbegin.o crtend.o"
   ;;
+*-*-linux-ohos*)
+  tmake_file="$tmake_file t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-gld t-slibgcc-elf-ver t-linux"
+  tmake_file="$tmake_file t-ohos-crtstuff"
+  extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
+  ;;
 *-*-linux* | frv-*-*linux* | *-*-kfreebsd*-gnu | *-*-gnu* | *-*-kopensolaris*-gnu | *-*-uclinuxfdpiceabi)
   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip t-slibgcc t-slibgcc-gld t-slibgcc-elf-ver t-linux"
   extra_parts="crtbegin.o crtbeginS.o crtbeginT.o crtend.o crtendS.o"
@@ -435,6 +440,16 @@ aarch64*-*-fuchsia*)
 	tmake_file="${tmake_file} ${cpu_type}/t-softfp t-softfp"
 	tmake_file="${tmake_file} t-dfprules"
 	;;
+aarch64*-*-linux-ohos*)
+	extra_parts="$extra_parts crtfastmath.o"
+	md_unwind_def_header=aarch64/aarch64-unwind-def.h
+	md_unwind_header=aarch64/linux-unwind.h
+	tmake_file="${tmake_file} ${cpu_type}/t-aarch64"
+	tmake_file="${tmake_file} ${cpu_type}/t-lse t-slibgcc-libgcc"
+	tmake_file="${tmake_file} ${cpu_type}/t-softfp t-softfp"
+	tmake_file="${tmake_file} t-dfprules"
+	tmake_file="${tmake_file} t-ohos-crtstuff"
+	;;
 aarch64*-*-linux*)
 	extra_parts="$extra_parts crtfastmath.o"
 	md_unwind_def_header=aarch64/aarch64-unwind-def.h
@@ -552,6 +567,14 @@ arm*-*-netbsdelf*)
 	    ;;
 	esac
 	;;
+arm*-*-linux-ohos*)
+	tmake_file="${tmake_file} arm/t-arm t-fixedpoint-gnu-prefix"
+	tmake_file="${tmake_file} arm/t-elf arm/t-bpabi arm/t-linux-eabi t-slibgcc-libgcc"
+	tm_file="$tm_file arm/bpabi-lib.h"
+	unwind_header=config/arm/unwind-arm.h
+	tmake_file="$tmake_file t-softfp-sfdf t-softfp-excl arm/t-softfp t-softfp"
+	tmake_file="${tmake_file} t-ohos-crtstuff"
+	;;
 arm*-*-linux* | arm*-*-uclinuxfdpiceabi)
 	tmake_file="${tmake_file} arm/t-arm t-fixedpoint-gnu-prefix t-crtfm"
 	tmake_file="${tmake_file} arm/t-elf arm/t-bpabi arm/t-linux-eabi t-slibgcc-libgcc"
@@ -811,6 +834,22 @@ i[34567]86-*-gnu*)
 	tm_file="${tm_file} i386/elf-lib.h"
 	md_unwind_header=i386/gnu-unwind.h
 	;;
+i[34567]86-*-linux-ohos*)
+	extra_parts="$extra_parts crtprec32.o crtprec64.o crtprec80.o crtfastmath.o"
+	tmake_file="${tmake_file} i386/t-crtpc t-crtfm i386/t-crtstuff t-dfprules"
+	tmake_file="${tmake_file} t-ohos-crtstuff"
+	tm_file="${tm_file} i386/elf-lib.h"
+	md_unwind_header=i386/linux-unwind.h
+	tmake_file="${tmake_file} i386/t-heap-trampoline"
+	;;
+x86_64-*-linux-ohos*)
+	extra_parts="$extra_parts crtprec32.o crtprec64.o crtprec80.o crtfastmath.o"
+	tmake_file="${tmake_file} i386/t-crtpc t-crtfm i386/t-crtstuff t-dfprules"
+	tmake_file="${tmake_file} t-ohos-crtstuff"
+	tm_file="${tm_file} i386/elf-lib.h"
+	md_unwind_header=i386/linux-unwind.h
+	tmake_file="${tmake_file} i386/t-heap-trampoline"
+	;;
 x86_64-*-linux*)
 	extra_parts="$extra_parts crtprec32.o crtprec64.o crtprec80.o crtfastmath.o"
 	tmake_file="${tmake_file} i386/t-crtpc t-crtfm i386/t-crtstuff t-dfprules"
@@ -1063,6 +1102,11 @@ microblaze*-*-rtems*)
 	;;
 mips*-*-netbsd*)			# NetBSD/mips, either endian.
 	;;
+mips*-*-linux-ohos*)			# Linux OHOS MIPS, either endian.
+	tmake_file="${tmake_file} mips/t-mips16 t-slibgcc-libgcc"
+	tmake_file="${tmake_file} t-ohos-crtstuff"
+	md_unwind_header=mips/linux-unwind.h
+	;;
 mips*-*-linux*)				# Linux MIPS, either endian.
 	extra_parts="$extra_parts crtfastmath.o"
 	tmake_file="${tmake_file} t-crtfm"
@@ -1349,6 +1393,12 @@ pru-*-*)
 	tmake_file="${tmake_file} t-softfp-sfdf t-softfp-excl t-softfp t-gnu-prefix pru/t-pru"
 	tm_file="$tm_file pru/pru-abi.h"
 	;;
+riscv*-*-linux-ohos*)
+	tmake_file="${tmake_file} riscv/t-crtstuff riscv/t-softfp${host_address} t-softfp riscv/t-elf riscv/t-elf${host_address} t-slibgcc-libgcc"
+	extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o crtendS.o crtbeginT.o"
+	tmake_file="${tmake_file} t-ohos-crtstuff"
+	md_unwind_header=riscv/linux-unwind.h
+	;;
 riscv*-*-linux*)
 	tmake_file="${tmake_file} riscv/t-crtstuff riscv/t-softfp${host_address} t-softfp riscv/t-elf riscv/t-elf${host_address} t-slibgcc-libgcc"
 	extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o crtendS.o crtbeginT.o"
@@ -1612,7 +1662,7 @@ i[34567]86-*-* | x86_64-*-*)
 esac
 
 case ${host} in
-i[34567]86-*-linux* | x86_64-*-linux*)
+i[34567]86-*-linux* | i[34567]86-*-linux-ohos* | x86_64-*-linux* | x86_64-*-linux-ohos*)
 	# Provide backward binary compatibility for 64bit Linux/x86.
 	if test "${host_address}" = 64; then
 		tmake_file="${tmake_file} i386/${host_address}/t-softfp-compat"
diff --git a/libgcc/config/t-ohos-crtstuff b/libgcc/config/t-ohos-crtstuff
new file mode 100644
index 000000000..8baa33873
--- /dev/null
+++ b/libgcc/config/t-ohos-crtstuff
@@ -0,0 +1,4 @@
+# OHOS uses PIE by default, so all crt* files need to be compiled with PIC
+# to avoid R_X86_64_32 relocation issues against hidden symbols like __TMC_END__
+CRTSTUFF_T_CFLAGS = $(PICFLAG)
+CRTSTUFF_T_CFLAGS_S = $(CRTSTUFF_T_CFLAGS) $(PICFLAG)
diff --git a/libgcc/gthr-posix.h b/libgcc/gthr-posix.h
index eeb84f32d..1b5cd8bc5 100644
--- a/libgcc/gthr-posix.h
+++ b/libgcc/gthr-posix.h
@@ -129,7 +129,7 @@ __gthrw(pthread_join)
 __gthrw(pthread_equal)
 __gthrw(pthread_self)
 __gthrw(pthread_detach)
-#ifndef __BIONIC__
+#if !defined(__BIONIC__) && !defined(__OHOS__)
 __gthrw(pthread_cancel)
 #endif
 __gthrw(sched_yield)
@@ -275,7 +275,7 @@ __gthrw2(__gthrw_(__pthread_key_create),
 	 __pthread_key_create,
 	 pthread_key_create)
 # define GTHR_ACTIVE_PROXY	__gthrw_(__pthread_key_create)
-#elif defined (__BIONIC__)
+#elif defined (__BIONIC__) || defined (__OHOS__)
 # define GTHR_ACTIVE_PROXY	__gthrw_(pthread_create)
 #else
 # define GTHR_ACTIVE_PROXY	__gthrw_(pthread_cancel)
diff --git a/libstdc++-v3/configure.host b/libstdc++-v3/configure.host
index 253e5a9ad..b2af509fb 100644
--- a/libstdc++-v3/configure.host
+++ b/libstdc++-v3/configure.host
@@ -268,6 +268,9 @@ case "${host_os}" in
   linux-musl*)
     os_include_dir="os/generic"
     ;;
+  linux-ohos*)
+    os_include_dir="os/generic"
+    ;;
   gnu* | linux* | kfreebsd*-gnu | uclinux*)
     if [ "$uclibc" = "yes" ]; then
       os_include_dir="os/uclibc"
-- 
2.34.1

